freqtrade/rpc/api_server/api_ws.py

```diff
@@ -0,0 +1,142 @@
+                                                 WSRequestSchema, WSWhitelistMessage)
+from freqtrade.rpc.rpc import RPC
+
+
+logger = logging.getLogger(__name__)
+
+# Private router, protected by API Key authentication
+router = APIRouter()
+
+
+async def is_websocket_alive(ws: WebSocket) -> bool:
+    """
+    Check if a FastAPI Websocket is still open
+    """
+    if (
+        ws.application_state == WebSocketState.CONNECTED and
+        ws.client_state == WebSocketState.CONNECTED
+    ):
+        return True
+    return False
+
+
+async def _process_consumer_request(
+    request: Dict[str, Any],
+    channel: WebSocketChannel,
+    rpc: RPC
+):
+    """
+    Validate and handle a request from a websocket consumer
+    """
+    # Validate the request, makes sure it matches the schema
+    try:
+        websocket_request = WSRequestSchema.parse_obj(request)
+    except ValidationError as e:
+        logger.error(f"Invalid request from {channel}: {e}")
+        return
+
+    type, data = websocket_request.type, websocket_request.data
+    response: WSMessageSchema
+
+    logger.debug(f"Request of type {type} from {channel}")
+
+    # If we have a request of type SUBSCRIBE, set the topics in this channel
+    if type == RPCRequestType.SUBSCRIBE:
+        # If the request is empty, do nothing
+        if not data:
+            return
+
+        # If all topics passed are a valid RPCMessageType, set subscriptions on channel
+        if all([any(x.value == topic for x in RPCMessageType) for topic in data]):
+            channel.set_subscriptions(data)
+
+        # We don't send a response for subscriptions
+        return
+
+    elif type == RPCRequestType.WHITELIST:
+        # Get whitelist
+        whitelist = rpc._ws_request_whitelist()
+
+        # Format response
+        response = WSWhitelistMessage(data=whitelist)
+        # Send it back
+        await channel.send(response.dict(exclude_none=True))
+
+    elif type == RPCRequestType.ANALYZED_DF:
+        limit = None
+
+        if data:
+            # Limit the amount of candles per dataframe to 'limit' or 1500
+            limit = max(data.get('limit', 1500), 1500)
+
+        # They requested the full historical analyzed dataframes
+        analyzed_df = rpc._ws_request_analyzed_df(limit)
+
+        # For every dataframe, send as a separate message
+        for _, message in analyzed_df.items():
+            response = WSAnalyzedDFMessage(data=message)
+            await channel.send(response.dict(exclude_none=True))
+
+
+@router.websocket("/message/ws")
+async def message_endpoint(
+    ws: WebSocket,
+    rpc: RPC = Depends(get_rpc),
+    channel_manager=Depends(get_channel_manager),
+    token: str = Depends(validate_ws_token)
+):
+    """
+    Message WebSocket endpoint, facilitates sending RPC messages
+    """
+    try:
+        # TODO:
+        # Return a channel ID, pass that instead of ws to the rest of the methods
```

Comment:
Can you explain this todo?

i don't think you can easily replace passing `ws` around - you'll need the reference to it to call methods ... so i don't really understand what would be behind this.

Reply:
So this was originally because I wanted to just use the channel manager to interact with the channels such as send messages and such, and it would align the interface where every method would take first the channel ID as a parameter. However we ended up just using the channels directly in the endpoint so this is irrelevant now.